website/addons/owncloud/model.py

```diff
@@ -0,0 +1,153 @@
+
+    @property
+    def folder_path(self):
+        return self.folder_id
+
+    @property
+    def folder_name(self):
+        return self.fetch_folder_name()
+
+    def set_folder(self, folder, auth=None):
+        if folder == '/ (Full ownCloud)':
+            folder = '/'
+        self.folder_id = folder
+        self.save()
+        if auth:
+            self.nodelogger.log(action='folder_selected', save=True)
+
+    def fetch_folder_name(self):
+        if self.folder_id == '/':
+            return '/ (Full ownCloud)'
+        return self.folder_id
+
+    def clear_settings(self):
+        self.folder_id = None
+
+    def deauthorize(self, auth=None, add_log=True):
+        """Remove user authorization from this node and log the event."""
+        self.clear_settings()
+        self.clear_auth()
+        if add_log and auth:
+            self.nodelogger.log(action='node_deauthorized', save=True)
+
+    def serialize_waterbutler_credentials(self):
+        if not self.has_auth:
+            raise exceptions.AddonError('Addon is not authorized')
+        provider = OwnCloudProvider(self.external_account)
+        return {'host': provider.host,
+                'username': provider.username,
+                'password': provider.password
+                }
+
+    def serialize_waterbutler_settings(self):
+        if not self.folder_id:
+            raise exceptions.AddonError('ownCloud is not configured')
+        return {'folder': self.folder_id,
+                'verify_ssl': USE_SSL}
+
+    def create_waterbutler_log(self, auth, action, metadata):
+        url = self.owner.web_url_for('addon_view_or_download_file', path=metadata['path'], provider='owncloud')
+        self.owner.add_log(
+            'owncloud_{0}'.format(action),
+            auth=auth,
+            params={
+                'project': self.owner.parent_id,
+                'node': self.owner._id,
+                'folder': self.folder_id,
+                'filename': metadata['materialized'].strip('/'),
+                'urls': {
+                    'view': url,
+                    'download': url + '?action=download'
+                },
+            },
+        )
+
+    def after_delete(self, node, user):
+        self.deauthorize(Auth(user=user), add_log=True)
+        self.save()
+
+    def on_delete(self):
+        self.deauthorize(add_log=False)
+        self.save()
+
+    def get_folders(self, path='/', folder_id=None):
+
+        provider = OwnCloudProvider(account=self.external_account)
+
+        c = OwnCloudClient(provider.host, verify_certs=settings.USE_SSL)
+        c.login(provider.username, provider.password)
+
+        ret = []
+        for item in c.list(path):
+            if item.file_type is 'dir':
+                ret.append({
+                    'addon': 'owncloud',
+                    'path': item.path,
+                    'kind': 'folder',
+                    'id': item.path,
+                    'name': item.path.strip('/').split('/')[-1],
```

Comment:
Ah, for a minute there, I thought you meant using kwargs via the above comment. In this case, I used the conversion from path to name. If that's incorrect, let me know.